axum/CHANGELOG.md

```diff
@@ -7,56 +7,308 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-  for `/foo`, axum would redirect calls to `/foo/` to `/foo` (or vice versa for
-  `/foo/`). That is no longer supported and such requests will now be sent to
-  the fallback. Consider using `axum_extra::routing::RouterExt::route_with_tsr`
-  if you want the old behavior ([#1119])
-- **fixed:** If `WebSocketUpgrade` cannot upgrade the connection it will return a
-  `WebSocketUpgradeRejection::ConnectionNotUpgradable` rejection ([#1135])
-- **changed:** `WebSocketUpgradeRejection` has a new variant `ConnectionNotUpgradable`
-  variant ([#1135])
+
+  For example:
+
+  ```rust
+  use axum::{Router, routing::get};
+
+  let app = Router::new()
+      // this used to not be allowed but now just works
+      .route("/foo/*rest", get(foo))
+      .route("/foo/bar", get(bar));
+
+  async fn foo() {}
+
+  async fn bar() {}
+  ```
+
+- **breaking:** Trailing slash redirects have been removed. Previously if you
+  added a route for `/foo`, axum would redirect calls to `/foo/` to `/foo` (or
+  vice versa for `/foo/`). That is no longer supported and such requests will
+  now be sent to the fallback. Consider using
+  `axum_extra::routing::RouterExt::route_with_tsr` if you want the old behavior
+  ([#1119])
+
+  For example:
+
+  ```rust
+  use axum::{Router, routing::get};
+
+  let app = Router::new()
+      // a request to `GET /foo/` will now get `404 Not Found`
+      // whereas in 0.5 axum would redirect to `/foo`
+      //
+      // same goes the other way if you had the route `/foo/`
+      // axum will no longer redirect from `/foo` to `/foo/`
+      .route("/foo", get(handler));
+
+  async fn handler() {}
+  ```
+
+- **breaking:** `Router::fallback` now only accepts `Handler`s (similarly to
+  what `get`, `post`, etc accept) ([#1155])
+
+  This fallback on 0.5:
+
+  ```rust
+  use axum::{Router, handler::Handler};
+
+  let app = Router::new().fallback(fallback.into_service());
+
+  async fn fallback() {}
+  ```
+
+  Becomes this in 0.6
+
+  ```rust
+  use axum::Router;
+
+  let app = Router::new().fallback(fallback);
+
+  async fn fallback() {}
+  ```
+
+- **added:** `Router::fallback_service` for setting any `Service` as the
+  fallback ([#1155])
+- **breaking:** Allow `Error: Into<Infallible>` for `Route::{layer, route_layer}` ([#924])
+- **breaking:** `MethodRouter` now panics on overlapping routes ([#1102])
 - **added** Implement `TryFrom<http:: Method>` for `MethodFilter` and use new `NoMatchingMethodFilter` error in case of failure ([#1130])
-- **added:** Support running extractors from `middleware::from_fn` functions ([#1088])
-- **added:** Added `debug_handler` which is an attribute macro that improves
-  type errors when applied to handler function. It is re-exported from
-  `axum-macros`
+- **breaking:** `Router::route` now only accepts `MethodRouter`s created with
+  `get`, `post`, etc ([#1155])
+- **added:** `Router::route_service` for routing to arbitrary `Service`s ([#1155])
```

Comment:
The first item is kind of missing "instead of arbitrary services", also I think it would make sense to merge these two into one since they're so closely related.